base/iterators.jl

```diff
@@ -780,16 +782,17 @@ length(P::ProductIterator) = prod(size(P))
 
 IteratorEltype(::Type{ProductIterator{Tuple{}}}) = HasEltype()
 IteratorEltype(::Type{ProductIterator{Tuple{I}}}) where {I} = IteratorEltype(I)
+
 function IteratorEltype(::Type{ProductIterator{T}}) where {T<:Tuple}
-    I = tuple_type_head(T)
-    P = ProductIterator{tuple_type_tail(T)}
-    IteratorEltype(I) == EltypeUnknown() ? EltypeUnknown() : IteratorEltype(P)
+    E = ntuple(n -> IteratorEltype(fieldtype(T, n)), _counttuple(T))
+    any(I -> I == EltypeUnknown(), E) && return EltypeUnknown()
+    return E[end]
 end
 
-eltype(::Type{<:ProductIterator{I}}) where {I} = _prod_eltype(I)
+eltype(::Type{<:ProductIterator{I}}) where {I} = @isdefined(I) ? _prod_eltype(I) : Any
```

Comment:
does using `@isdefined(I)` here help the compiler when calling `_prod_eltype(I)`? I imagine this is for the case when `I` isn't inferrable?

Reply:
It was supposed to help with subtypes of `ProductIterator`, but since it doesn't have subtypes, I think I can just eliminate that:
```suggestion
eltype(::Type{ProductIterator{I}}) where {I} = _prod_eltype(I)
```